realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -2474,3 +2486,480 @@ void TemperatureDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWra
+                            || (image_publisher.first.getNumSubscribers() != 0);
+        if (has_subscribers) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// This callback is implemented assuming that ROS service callbacks are serialized,
+// which seems to be true for this node at the moment of writing.
+// Refer to http://wiki.ros.org/nodelet#Threading_Model to find out how to verify it.
+bool BaseRealSenseNode::nomagicGetLatestFrameCallback(stream_index_pair stream, bool is_aligned_depth,
+                                                      GetLatestFrame::Request& request,
+                                                      GetLatestFrame::Response& response)
+{
+    response.request_timestamp = nomagicGetUnixTimestamp();
+    ROS_INFO("[NOMAGIC] get_latest_frame: stream=%s_%d aligned_depth=%d",
+             rs2_stream_to_string(stream.first), stream.second, is_aligned_depth);
+    Clock clock;
+
+    // If nomagic_lazy_filtering, queue will have == 1 frameset
+    // Otherwise, there will be >= 1 frameset
+    clock.restart();
+    auto queue = nomagicGetNonEmptyFramesetQueue();
+    response.wait_for_frames_duration = clock.getElapsedSecs();
+    ROS_INFO("[NOMAGIC] Queue contains %lu frames:", queue.size());
+
+    for (auto&& frameset : queue) {
+        ROS_INFO("[NOMAGIC] %s", nomagicFramesetDescriptionString(frameset).c_str());
+    }
+
+    // Here we're assuming that the timestamps are in the same domain.
+    // If that's no the case, it will be visible in the logs printed above
+    double actual_timespan = (queue.back().get_timestamp() - queue.front().get_timestamp()) / 1000.0;
+    double expected_timespan =
+            static_cast<double>(queue.size() - 1) / (is_aligned_depth ? _fps[DEPTH] : _fps[stream]);
+    ROS_INFO("[NOMAGIC] Buffer time span: %fs (expected: %fs)", actual_timespan, expected_timespan);
+
+    // Parameter stream means usually the type of the returned frame (color/depth/infra1/infra2)
+    // However, when is_aligned_depth, we will return depth aligned to {stream}
+
+    rs2::frame final_frame;
+    // If we're not requested for aligned depth,
+    // RGB and Infra frames don't need any processing:
+    if (!is_aligned_depth && stream != DEPTH) {
+        rs2::frameset frameset = nomagic_frameset_queue.back();
+        final_frame = nomagicFramesetToFrame(stream, frameset);
+    }
+    else {
+        // If nomagic_lazy_filtering, we need to clean the temporal filter state
+        // as it may contain old data from the previous request.
+        if (nomagic_lazy_filtering) {
+            clock.restart();
+            nomagicResetTemporalFilter();
+            response.reset_temporal_filter_duration = clock.getElapsedSecs();
+        }
+
+        // If nomagic_lazy_filtering, we apply filters for all the frames we have kept.
+        // Otherwise, the queue will be of size == 1 and filtering will be done only for the most recent frame
```

Comment:
Does it mean that is nomagic_lazy_filtering is false, not temporal filtering will take place, even if I have it set up in the config file? If so, it would be confusing.